README.md
# Musy API Server

Simple python3 API server to create a mix of music tracks for LMS.

## Musly

This service uses the [Musly audio music similarity library](https://github.com/CDrummond/musly)
This library needs to be compiled, and `config.json` (from this API project)
updated to store the location of this library. This repo contains pre-built
built versions for:

1. Fedora 64-bit
2. Raspbian Buster 32-bit

**macOS**

Instructions, and binaries, for Musly on macOS can be found [here](https://github.com/AF-1/sobras/tree/main/lms-musly-server_on_macos)

## Analysing Tracks

Before starting the server your music needs to be analysed with Musly. This is
accomplished via:

```
./musly-server.py --analyse /path/to/music/folder
```

This takes between 1 and 2 hours to process 25k tracks. The process analyses
tracks, adds them to Musly, intialises Musly's 'jukebox' style with 1000 random
tracks, and extracts certain tags. If re-run new tracks will be added, and old
(non-existant) will be removed. Pass `--keep-old` to keep these old tracks.

To analyse the Musly path stored in the config file, the following shortcut can
be used:

```
./musly-server.py --analyse m
```

### CUE files

If the analysis locates a music file with a similarly named CUE file (e.g.
`artist/album/album name.flac` and `artist/album/album name.cue`) then it will
read the track listing from the LMS db file and use `ffmpeg` to split the
music file into temporary 128kbps MP3 files for analysis. The files are removed
once analysis is complete.


## Testing Analysis

Musly has a [bug](https://github.com/dominikschnitzer/musly/issues/43) where
sometimes it gives the same similarity to all tracks, which will obviously break
the API this script is designed for. To test if the analysis is correct you can
run the script in test mode:

```
./musly-server.py --log-level INFO --test
```

This will query Musly for the 50 most similar tracks to the 1st analysed track.
It then checks that there are different similarities, and if not an error
message is shown.

If this script states that there is an error you can try simply removing the
jukebox file and re-runnnig this script - it will recreate the jukebox with
random tracks. If this keeps failing it might be better to adjust the
`styletracks` config item, delete the jukebox, and test again.

Alternatively, you can have this script run until it receives valid similarities
from Musly. In this mode the script will test the jukebox, remove it if the test
fails, recreate jukebox, test the jukebox, ...

```
./musly-server.py --log-level INFO --test --repeat
```

## Similarity API 

The API server can be installed as a Systemd service, or started manually:

```
./musly-server.py
```

...when the service starts, it will confirm that the number of tracks in its
SQLite database is the same as the number in the 'Musly jukebox'. If the
number differs, the jukebox is recreated.

Only 1 API is currently supported:

```
http://HOST:11000/api/similar?track=/path/of/track&track=/path/of/another/track&count=10&filtergenre=1&min=30&max=600&norepart=15&norepalb=25&filterxmas=1
```
...this will get 10 similar tracks to those supplied.

If `filtergenre=1` is supplied then only tracks whose genre matches a
pre-configured set of genres (mapped from seed tracks) will be used. e.g. if
`["Heavy Metal", "Metal", "Power Metal"]` is defined in the config, and a seed
tack's genre has `Metal` then only tracks with one of these 3 genres will be
considered.

If `filterxmas=1` is supplied, then tracks with 'Christmas' or 'Xmas' in their
genres will be excluded - unless it is December.

`min` and `max` can be used to set the minimum, and maximum, duration (in
seconds) of tracks to be considered.

`norepart` specifies the number of tracks where an artist should not be
repeated. This is not a hard-limit, as if there are too few candidates then
repeats can happen.

`norepalb` specifies the number of tracks where an album should not be
repeated. This does not aply to 'Various Artist' albums. This is also not a
hard-limit, as if there are too few candidates then repeats can happen.

`previous` may be used to list tracks currently in the play queue. This
parameter, like `track`, may be repeated multiple times. These tracks will be
used to filter chosen tracks on artist, or album, to prevent duplicates.

`maxsim` (range 0..100) can be used to set the maximum similarity factor. A
factor of 0 would imply the track is identical, 100 completely different. 75 is
the default value.

`shuffle` if set to `1` will cause extra tracks to be located, this list
shuffled, and then the desired `count` tracks taken from this shuffled list.

The API will use Musly to get the similairt between all tracks and each seed
track, and sort this by similarity (most similar first). Initally the API will
ignore Musly tracks from the same artist or album of the seed tracks (and any
previous in the list, any albums from the (e.g.) 25 `previous` tracks, or albums
from the last (e.g.) 15 `previous` tracks). If, because of this filtering,
there are less than the requested amount then the highest similarity tracks
from the filtered-out lists are chosen.

Metadata for tracks is stored in an SQLite database, this has an `ignore` column
which if set to `1` will cause the API to not use this track if it is returned
as a similar track by Musly. In this way you can exclude specific tracks from
being added to mixes - but if they are already in the queue, then they can still
be used as seed tracks.

This API is intended to be used by [LMS Music Similarity Plugin](https://github.com/CDrummond/lms-musicsimilarity)

Genres are configured via the `genres` section of `config.json`, using the
following syntax:

```
{
 "genres:[
  [ "Rock", "Hard Rock", "Metal" ],
  [ "Pop", "Dance", "R&B"]
 ]
}
```

If a seed track has `Hard Rock` as its genre, then only tracks with `Rock`,
`Hard Rock`, or `Metal` will be allowed. If a seed track has a genre that is not
listed here then any track returned by Musly, that does not contain any genre
listed here, will be considered acceptable. Therefore, if seed is `Pop` then
a `Hard Rock` track would not be considered.

### HTTP Post

Alternatively, the API may be accessed via a HTTP POST call. To do this, the
params of the call are passed as a JSON object. eg.

```
{
 "track":["/path/trackA.mp3", "/path/trackB.mp3"],
 "filtergenre":1,
 "count":10
}
```

This is the method that is used by the LMS plugin.

## Configuration

The sever reads its configuration from a JSON file (default name is
`config.json`). This has the following format:

```
{
 "libmusly":"lib/x86-64/fedora/libmusly.so",
 "paths":{
  "db":"/home/user/.local/share/musly/",
  "musly":"/home/Music/",
  "lms":"/media/Music/",
  "tmp":"/tmp/"
 },
 "lmsdb":"/path/to/lms/Cache/library.db",
 "genres":[
  ["Alternative Rock", "Classic Rock", "Folk/Rock", "Hard Rock", "Indie Rock", "Punk Rock", "Rock"],
  ["Dance", "Disco", "Hip-Hop", "Pop", "Pop/Folk", "Pop/Rock", "R&B", "Reggae", "Soul", "Trance"],
  ["Gothic Metal", "Heavy Metal", "Power Metal", "Progressive Metal", "Progressive Rock", "Symphonic Metal", "Symphonic Power Metal"]
 ],
 "ignoregenre":["Artist"],
 "normalize":{
  "artist":["feet", "ft", "featuring"],
  "album":["deluxe edition", "remastered"],
  "title"["demo", "radio edit"]
 },
 "port":10000,
 "host":"0.0.0.0",
 "threads":8,
 "styletracks":1000,
 "styletracksmethod":"genres",
 "extractstart":-48,
 "extractlen":30
}
```

* `libmusly` should contain the path the musy shared library - path is relative
to `musly-server.py`
* `paths.db` should be the path where the SQLite and jukebox files created by
this app can be written.
* `paths.musly` should be the path where Musly can access your music files. This
can be different to `path.lms` if you are running analysis on a different
machine to where you would run the script as the API server. This script will
only store the paths relative to this location - eg. `paths.musly=/home/music/`
then `/home/music/A/b.mp3` will be stored as `A/b.mp3`.
* `paths.lms` should be the path where LMS accesses your music files. The API
server will remove this path from API calls, so that it can look up tracks in
its database by their relative path.
* `paths.tmp` When analysing music, this script will create a temporary folder
to hold separate CUE file tracks. The path passed here needs to be writable.
This config item is only used for analysis.
* `lmsdb` During analysis, this script will also analyse individual CUE tracks.
To do this it needs access to the LMS database file to know the position of each
track, etc. This config item should hole the path to the LMS database file. This
is only required for analysis, and only if you have CUE files. `ffmpeg` is
required to split tracks.
* `genres` This is as described above.
* `ignoregenre` List of artists where genre filtering (excluding christmas)
should be ignored. To apply to all artists, use '*' - e.g. `"ignoregenre":"*"`
* `normalize.artist` List of strings to split artist names, e.g. "A ft. B"
becomes "A" (periods are automatically removed).
* `normalize.album` List of strings to remove from album names.
* `normalize.title` List of strings to remove from titles.
* `port` This is the port number the API is accessible on.
* `host` IP addres on which the API will listen on. Use `0.0.0.0` to listen on
all interfaces on your network.
* `threads` Number of threads to use during analysis phase. This controls how
many calls to `ffmpeg` are made concurrently, and how many concurrent tracks
Musly is asked to analyse. Defaults to CPU count, if not set.
* `styletracks` A  subset of tracks is passed to Musly's `setmusicstyle`
function, by default 1000 random tracks is chosen. This config item can be used
to alter this. Note, however, the larger the number here the longer it takes to
for this call to complete. As a rough guide it takes ~1min per 1000 tracks.
If you change this config item after the jukebox is written you will need to
delete the jukebox file and restart the server.
* `styletracksmethod` configures how tracks are chosen for styletracks. If
set to `genres` (which is the default if not set) then the meta-data db is
queried for how many track each genre has and tracks are chosen for each of
these genres based upon the percentage of tracks in a genre. If set to `albums`
then at least one track from each album is used. If set to aything else then
random tracks are chosen.
* `extractlen` The maximum length in seconds of the file to decode. If zero
or greater than the file length, then the whole file will be decoded. Note,
however, that only a maximum of 5 minutes is used for analysis.
* `extractstart` The starting position in seconds of the excerpt to decode. If
zero, decoding starts at the beginning. If negative, the excerpt is centered in
the file, but starts at -`extractstart` the latest. If positive and
`extractstart`+`extractlen` exceeds the file length, then the excerpt is taken
from the end of the file.


## Ignoring artists, albums, etc.

To mark certains items as 'ignored' (i.e. so that they are not added to mixes),
create a text file where each line contains the unique path, e.g.:

```
AC-DC/Power Up/
The Police/
```

Then call:

```
./update-db.py --db musly.db --ignore ignore.txt
```

This sets the `ignore` column to 1 for all items whose file starts with one of
the listed lines.

Setting a track's `ignore` to `1` will exclude tracks from being added to
mixes - but if they are already in the queue, then they can still be used as
seed tracks.


## Credits

`lib/musly.py` (which is used as a python interface to the Musly library) is
taken, and modified, from [Musly Integration for the Logitech Media Server](https://www.nexus0.net/pub/sw/lmsmusly)


ChangeLog
0.0.3
-----
 1. Change default 'extractlen' to 30, to match Musly recommended value.
 2. Make extract start and length configurable.
 3. Allow configuring no repeat artist/album/track.
 4. Fix parsing of 1/0 parameters.
 5. Display track number, total, and percentage during analysis.
 6. Add a test mode to test musly similarities.
 7. Expand /api/dump to allow usage in MusicSimilarity LMS plugin.
 8. When choosing style tracks, get random tracks from each genre.
 9. Add setting to control how style tracks are chosen.
10. Don't use musly_findmin as it seems to corrupt results, instead just sort
    tracks by similarity before processing.
11. Build musly library using code from https://github.com/CDrummond/musly
12. Fix multi-threaded analysis - use separate process for each track.

0.0.2
-----
 1. Add version string.
 2. Place ZIP contents into musly-server folder.
 3. Add 'minsim' and 'maxsim' parameters to control the similarity range.
 4. Fix reading, and usage, of 'ignore' DB column.
 5. Remove 'id' column, and use rowid instead.
 6. Vacumm DB after removing tracks to reset rowids.
 7. Pass previous tracks to server and have server use this for filtering.
 8. Optionally (true by default) shuffle tracks.
 9. If an artist has multiple similar tracks, choose one at random.
10. When selecting tracks for musly jukebox style tracks, try to choose at least
    one random track in the range 60..300 seconds from each album.

0.0.1
-----
1. Initial release.


lib/app.py
#
# Analyse files with Musly, and provide an API to retrieve similar tracks
#
# Copyright (c) 2020-2021 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

import argparse
from datetime import datetime
import json
import logging
import math
import os
import random
import sqlite3
import urllib
from flask import Flask, abort, request
from . import cue, filters, metadata_db, musly

_LOGGER = logging.getLogger(__name__)

DEFAULT_TRACKS_TO_RETURN              = 5  # Number of tracks to return, if none specified
MIN_TRACKS_TO_RETURN                  = 5  # Min value for 'count' parameter
MAX_TRACKS_TO_RETURN                  = 50 # Max value for 'count' parameter
DEFAULT_NUM_PREV_TRACKS_FILTER_ARTIST = 15 # Try to ensure artist is not in previous N tracks
DEFAULT_NUM_PREV_TRACKS_FILTER_ALBUM  = 25 # Try to ensure album is not in previous N tracks
NUM_SIMILAR_TRACKS_FACTOR             = 25 # Request count*NUM_SIMILAR_TRACKS_FACTOR from musly
SHUFFLE_FACTOR                        = 3  # How many (shuffle_factor*count) tracks to shuffle?


class MuslyApp(Flask):
    def init(self, args, mus, app_config, jukebox_path):
        _LOGGER.debug('Start server')
        self.app_config = app_config
        self.mus = mus
        
        flask_logging = logging.getLogger('werkzeug')
        flask_logging.setLevel(args.log_level)
        flask_logging.disabled = 'DEBUG'!=args.log_level
        meta_db = metadata_db.MetadataDb(app_config)
        (paths, tracks) = self.mus.get_alltracks_db(meta_db.get_cursor())
        random.seed()
        ids = None

        # If we can, load musly from jukebox...
        if os.path.exists(jukebox_path):
            ids = self.mus.get_jukebox_from_file(jukebox_path)

        if ids==None or len(ids)!=len(tracks):
            _LOGGER.debug('Adding tracks from DB to musly')
            ids = mus.add_tracks(tracks, app_config['styletracks'], app_config['styletracksmethod'], meta_db)
            self.mus.write_jukebox(jukebox_path)

        meta_db.close()
        self.mta=musly.MuslyTracksAdded(paths, tracks, ids)

    def get_config(self):
        return self.app_config

    def get_musly(self):
        return self.mus

    def get_mta(self):
        return self.mta
    
musly_app = MuslyApp(__name__)


def get_value(params, key, defVal, isPost):
    if isPost:
        return params[key] if key in params else defVal
    return params[key][0] if key in params else defVal


def decode(url, root):
    u = urllib.parse.unquote(url)
    if u.startswith('file://'):
        u=u[7:]
    elif u.startswith('tmp://'):
        u=u[6:]
    if u.startswith(root):
        u=u[len(root):]
    return cue.convert_from_cue_path(u)


def genre_adjust(seed, entry, seed_genres, all_genres, match_all_genres):
    if match_all_genres:
        return 0.0
    if 'genres' not in seed:
        return 0.1
    if 'genres' not in entry:
        return 0.1
    if seed['genres'][0]==entry['genres'][0]:
        # Exact genre match
        return 0.0
    if (seed_genres is not None and entry['genres'][0] not in seed_genres) or \
       (seed_genres is None and all_genres is not None and entry['genres'][0] in all_genres):
        return 0.05
    # Genre in group
    return 0.025


@musly_app.route('/api/dump', methods=['GET', 'POST'])
def dump_api():
    isPost = False
    if request.method=='GET':
        params = request.args.to_dict(flat=False)
    else:
        isPost = True
        params = request.get_json()
        _LOGGER.debug('Request: %s' % json.dumps(params))

    if not params:
        abort(400)

    if not 'track' in params:
        abort(400)

    if len(params['track'])!=1:
        abort(400)

    mta = musly_app.get_mta()
    mus = musly_app.get_musly()
    cfg = musly_app.get_config()
    meta_db = metadata_db.MetadataDb(cfg)

    # Strip LMS root path from track path
    root = cfg['paths']['lms']

    track = decode(params['track'][0], root)
    _LOGGER.debug('S TRACK %s -> %s' % (params['track'][0], track))

    # Check that musly knows about this track
    track_id = -1
    try:
        track_id = mta.paths.index( track )
        if track_id<0:
            abort(404)
        fmt = get_value(params, 'format', '', isPost)
        txt = fmt=='text'
        txt_url = fmt=='text-url'
        match_artist = int(get_value(params, 'filterartist', '0', isPost))==1
        meta = meta_db.get_metadata(track_id+1) # IDs (rowid) in SQLite are 1.. musly is 0..

        all_genres = cfg['all_genres'] if 'all_genres' in cfg else None
        seed_genres=[]
        if 'genres' in meta and 'genres' in cfg:
            for genre in meta['genres']:
                for group in cfg['genres']:
                    if genre in group:
                        for cg in group:
                            if not cg in seed_genres:
                                seed_genres.append(cg)

        simtracks = mus.get_similars( mta.mtracks, mta.mtrackids, track_id )

        resp=[]
        prev_id=-1
        count = int(get_value(params, 'count', 1000, isPost))

        tracks=[]
        for simtrack in simtracks:
            if simtrack['id']==prev_id:
                break
            prev_id=simtrack['id']
            if math.isnan(simtrack['sim']):
                continue

            track = meta_db.get_metadata(simtrack['id']+1)
            if match_artist and track['artist'] != meta['artist']:
                continue
            match_all_genres = ('ignoregenre' in cfg) and (('*'==cfg['ignoregenre'][0]) or (meta is not None and meta['artist'] in cfg['ignoregenre']))
            sim = simtrack['sim'] + genre_adjust(meta, track, seed_genres, all_genres, match_all_genres)
            tracks.append({'path':mta.paths[simtrack['id']], 'sim':sim})

        tracks = sorted(tracks, key=lambda k: k['sim'])
        for track in tracks:
            _LOGGER.debug("%s %s" % (track['path'], track['sim']))
            if txt:
                resp.append("%s\t%f" % (track['path'], track['sim']))
            elif txt_url:
                resp.append(cue.convert_to_cue_url('%s%s' % (root, track['path'])))
            else:
                resp.append({'file':track['path'], 'sim':track['sim']})
            if len(resp)>=count:
                break
        if txt or txt_url:
            return '\n'.join(resp)
        else:
            return json.dumps(resp)
    except Exception as e:
        _LOGGER.error("EX:%s" % str(e))
        abort(404)


@musly_app.route('/api/similar', methods=['GET', 'POST'])
def similar_api():
    isPost = False
    if request.method=='GET':
        params = request.args.to_dict(flat=False)
    else:
        isPost = True
        params = request.get_json()
        _LOGGER.debug('Request: %s' % json.dumps(params))

    if not params:
        abort(400)

    if not 'track' in params:
        abort(400)

    count = int(get_value(params, 'count', DEFAULT_TRACKS_TO_RETURN, isPost))
    if count < MIN_TRACKS_TO_RETURN:
        count = MIN_TRACKS_TO_RETURN
    elif count > MAX_TRACKS_TO_RETURN:
        count = MAX_TRACKS_TO_RETURN

    match_genre = int(get_value(params, 'filtergenre', '0', isPost))==1
    shuffle = int(get_value(params, 'shuffle', '1', isPost))==1
    max_similarity = int(get_value(params, 'maxsim', 75, isPost))/100.0
    min_duration = int(get_value(params, 'min', 0, isPost))
    max_duration = int(get_value(params, 'max', 0, isPost))
    no_repeat_artist = int(get_value(params, 'norepart', 0, isPost))
    no_repeat_album = int(get_value(params, 'norepalb', 0, isPost))
    exclude_christmas = int(get_value(params, 'filterxmas', '0', isPost))==1 and datetime.now().month!=12

    if no_repeat_artist<0 or no_repeat_artist>200:
        no_repeat_artist = DEFAULT_NUM_PREV_TRACKS_FILTER_ARTIST
    if no_repeat_album<0 or no_repeat_album>200:
        no_repeat_album = DEFAULT_NUM_PREV_TRACKS_FILTER_ALBUM

    no_repeat_artist_or_album = no_repeat_album if no_repeat_album>no_repeat_artist else no_repeat_artist

    mta = musly_app.get_mta()
    mus = musly_app.get_musly()
    cfg = musly_app.get_config()
    meta_db = metadata_db.MetadataDb(cfg)

    # Strip LMS root path from track path
    root = cfg['paths']['lms']
    
    # Similar tracks
    similar_tracks=[]
    # Track IDs of similar tracks - used to avoid duplicates
    similar_track_ids=[]
    # Similar tracks ignored because of artist/album
    filtered_by_seeds_tracks=[]
    filtered_by_current_tracks=[]
    filtered_by_previous_tracks=[]
    current_titles=[]

    # Artist/album of seed tracks
    seed_metadata=[]
    track_id_seed_metadata={}
    seed_genres=[]
    all_genres = cfg['all_genres'] if 'all_genres' in cfg else None
    
    # Artist/album of chosen tracks
    current_metadata_keys={}
    current_metadata=[]

    if min_duration>0 or max_duration>0:
        _LOGGER.debug('Duration:%d .. %d' % (min_duration, max_duration))

    # Musly IDs of seed tracks
    track_ids = []
    for trk in params['track']:
        track = decode(trk, root)
        _LOGGER.debug('S TRACK %s -> %s' % (trk, track))

        # Check that musly knows about this track
        track_id = -1
        try:
            track_id = mta.paths.index( track )
            _LOGGER.debug('Get %d similar track(s) to %s, index: %d' % (count, track, track_id))
        except:
            pass
        if track_id is not None and track_id>=0:
            track_ids.append(track_id)
            meta = meta_db.get_metadata(track_id+1) # IDs (rowid) in SQLite are 1.. musly is 0..
            _LOGGER.debug('Seed %d metadata:%s' % (track_id, json.dumps(meta)))
            if meta is not None:
                seed_metadata.append(meta)
                track_id_seed_metadata[track_id]=meta
                # Get genres for this seed track - this takes its genres and gets any matching genres from config
                if 'genres' in meta and 'genres' in cfg:
                    for genre in meta['genres']:
                        for group in cfg['genres']:
                            if genre in group:
                                for cg in group:
                                    if not cg in seed_genres:
                                        seed_genres.append(cg)
                if 'title' in meta:
                    current_titles.append(meta['title'])
        else:
            _LOGGER.debug('Could not locate %s in DB' % track)

    previous_track_ids = []
    previous_metadata = [] # Ignore tracks with same meta-data, i.e. artist
    if 'previous' in params:
        for trk in params['previous']:
            track = decode(trk, root)
            _LOGGER.debug('I TRACK %s -> %s' % (trk, track))

            # Check that musly knows about this track
            track_id = -1
            try:
                track_id = mta.paths.index(track)
            except:
                pass
            if track_id is not None and track_id>=0:
                previous_track_ids.append(track_id)
                if len(previous_metadata)<no_repeat_artist_or_album:
                    meta = meta_db.get_metadata(track_id+1) # IDs (rowid) in SQLite are 1.. musly is 0..
                    if meta:
                        previous_metadata.append(meta)
                        if 'title' in meta:
                            current_titles.append(meta['title'])
            else:
                _LOGGER.debug('Could not locate %s in DB' % track)
        _LOGGER.debug('Have %d previous tracks' % len(previous_track_ids))

    if match_genre:
        _LOGGER.debug('Seed genres: %s' % seed_genres)

    similarity_count = int(count * SHUFFLE_FACTOR) if shuffle else count

    matched_artists={}
    for track_id in track_ids:
        match_all_genres = ('ignoregenre' in cfg) and (('*'==cfg['ignoregenre'][0]) or ((track_id in track_id_seed_metadata) and (track_id_seed_metadata[track_id]['artist'] in cfg['ignoregenre'])))

        # Query musly for similar tracks
        _LOGGER.debug('Query musly for similar tracks to index: %d' % track_id)
        simtracks = mus.get_similars( mta.mtracks, mta.mtrackids, track_id )

        accepted_tracks = 0
        for simtrack in simtracks:
            if math.isnan(simtrack['sim']):
                continue
            if (not simtrack['id'] in track_ids) and (not simtrack['id'] in previous_track_ids) and (not simtrack['id'] in similar_track_ids) and (simtrack['sim']>0.0) and (simtrack['sim']<=max_similarity):
                similar_track_ids.append(simtrack['id'])

                meta = meta_db.get_metadata(simtrack['id']+1) # IDs (rowid) in SQLite are 1.. musly is 0..
                if not meta:
                    _LOGGER.debug('DISCARD(not found) ID:%d Path:%s Similarity:%f' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim']))
                elif meta['ignore']:
                    _LOGGER.debug('DISCARD(ignore) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                elif (min_duration>0 or max_duration>0) and not filters.check_duration(min_duration, max_duration, meta):
                    _LOGGER.debug('DISCARD(duration) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                elif match_genre and not filters.genre_matches(cfg, seed_genres, meta):
                    _LOGGER.debug('DISCARD(genre) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                elif exclude_christmas and filters.is_christmas(meta):
                    _LOGGER.debug('DISCARD(xmas) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                else:
                    if filters.same_artist_or_album(seed_metadata, meta):
                        _LOGGER.debug('FILTERED(seeds) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                        filtered_by_seeds_tracks.append({'path':mta.paths[simtrack['id']], 'similarity':simtrack['sim']})
                    elif filters.same_artist_or_album(current_metadata, meta):
                        _LOGGER.debug('FILTERED(current) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                        filtered_by_current_tracks.append({'path':mta.paths[simtrack['id']], 'similarity':simtrack['sim']})
                        if meta['artist'] in matched_artists and simtrack['sim'] - matched_artists[meta['artist']]['similarity'] <= 0.2:
                            matched_artists[meta['artist']]['tracks'].append({'path':mta.paths[simtrack['id']], 'similarity':simtrack['sim']})
                    elif no_repeat_artist>0 and filters.same_artist_or_album(previous_metadata, meta, False, no_repeat_artist):
                        _LOGGER.debug('FILTERED(previous(artist)) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                        filtered_by_previous_tracks.append({'path':mta.paths[simtrack['id']], 'similarity':simtrack['sim']})
                    elif no_repeat_album>0 and filters.same_artist_or_album(previous_metadata, meta, True, no_repeat_album):
                        _LOGGER.debug('FILTERED(previous(album)) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                    elif filters.match_title(current_titles, meta):
                        _LOGGER.debug('FILTERED(title) ID:%d Path:%s Similarity:%f Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], json.dumps(meta)))
                        filtered_by_previous_tracks.append({'path':mta.paths[simtrack['id']], 'similarity':simtrack['sim']})
                    else:
                        key = '%s::%s::%s' % (meta['artist'], meta['album'], meta['albumartist'] if 'albumartist' in meta and meta['albumartist'] is not None else '')
                        if not key in current_metadata_keys:
                            current_metadata_keys[key]=1
                            current_metadata.append(meta)
                        sim = simtrack['sim'] + genre_adjust(seed_metadata, meta, seed_genres, all_genres, match_all_genres)

                        _LOGGER.debug('USABLE ID:%d Path:%s Similarity:%f AdjSim:%s Meta:%s' % (simtrack['id'], mta.paths[simtrack['id']], simtrack['sim'], sim, json.dumps(meta)))
                        similar_tracks.append({'path':mta.paths[simtrack['id']], 'similarity':sim})
                        # Keep list of all tracks of an artist, so that we can randomly select one => we don't always use the same one
                        matched_artists[meta['artist']]={'similarity':simtrack['sim'], 'tracks':[{'path':mta.paths[simtrack['id']], 'similarity':sim}], 'pos':len(similar_tracks)-1}
                        if 'title' in meta:
                            current_titles.append(meta['title'])
                        accepted_tracks += 1
                        if accepted_tracks>=similarity_count:
                            break

    # For each matched_artists randonly select a track...
    for matched in matched_artists:
        if len(matched_artists[matched]['tracks'])>1:
            _LOGGER.debug('Choosing random track for %s (%d tracks)' % (matched, len(matched_artists[matched]['tracks'])))
            sim = similar_tracks[matched_artists[matched]['pos']]['similarity']
            similar_tracks[matched_artists[matched]['pos']] = random.choice(matched_artists[matched]['tracks'])
            similar_tracks[matched_artists[matched]['pos']]['similarity'] = sim

    # Too few tracks? Add some from the filtered lists
    min_count = 2
    if len(similar_tracks)<min_count and len(filtered_by_previous_tracks)>0:
        _LOGGER.debug('Add some tracks from filtered_by_previous_tracks, %d/%d' % (len(similar_tracks), len(filtered_by_previous_tracks)))
        filtered_by_previous_tracks = sorted(filtered_by_previous_tracks, key=lambda k: k['similarity'])
        similar_tracks = similar_tracks + filtered_by_previous_tracks[:min_count-len(similar_tracks)]
    if len(similar_tracks)<min_count and len(filtered_by_current_tracks)>0:
        _LOGGER.debug('Add some tracks from filtered_by_current_tracks, %d/%d' % (len(similar_tracks), len(filtered_by_current_tracks)))
        filtered_by_current_tracks = sorted(filtered_by_current_tracks, key=lambda k: k['similarity'])
        similar_tracks = similar_tracks + filtered_by_current_tracks[:min_count-len(similar_tracks)]
    if len(similar_tracks)<min_count and len(filtered_by_seeds_tracks)>0:
        _LOGGER.debug('Add some tracks from filtered_by_seeds_tracks, %d/%d' % (len(similar_tracks), len(filtered_by_seeds_tracks)))
        filtered_by_seeds_tracks = sorted(filtered_by_seeds_tracks, key=lambda k: k['similarity'])
        similar_tracks = similar_tracks + filtered_by_seeds_tracks[:min_count-len(similar_tracks)]

    # Sort by similarity
    similar_tracks = sorted(similar_tracks, key=lambda k: k['similarity'])
    
    # Take top 'similarity_count' tracks
    similar_tracks = similar_tracks[:similarity_count]

    if shuffle:
        random.shuffle(similar_tracks)
        similar_tracks = similar_tracks[:count]

    track_list = []
    for track in similar_tracks:
        path = '%s%s' % (root, track['path'])
        track_list.append(cue.convert_to_cue_url(path))
        _LOGGER.debug('Path:%s %f' % (path, track['similarity']))

    meta_db.close()
    if get_value(params, 'format', '', isPost)=='text':
        return '\n'.join(track_list)
    else:
        return json.dumps(track_list)


def start_app(args, mus, config, jukebox_path):
    musly_app.init(args, mus, config, jukebox_path)
    _LOGGER.debug('Ready to process requests')
    musly_app.run(host=config['host'], port=config['port'])
